### ISSUE TITLE
High CPU usage when execute `soss ros1_to_ros2.yaml`

### ISSUE THREAD

**Issue by sparkskraps** [eProsima/ROS1-SH#6]
I let ROS2 and ROS1 communicate using SOSS, when i check CPU usage for ROSIS using top command, it shows like:
```
 PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                                                                            
10127 dbt       20   0  915644  38844  33800 S 100.0  0.2   1:48.40 soss                                                                                               
 9856 dbt       20   0  545804  55216   8284 S   0.3  0.3   0:02.93 rosmaster                                                                                          
 9867 dbt       20   0  336332  10540   9592 S   0.3  0.1   0:01.66 rosout                                                                                             
 9876 root      20   0       0      0      0 I   0.3  0.0   0:34.72 kworker/u8:1-ev 
 ....
```
When execute `soss  dds_to_ros2.yaml`  to let DDS and ROS2 communicate, the cpu usage is normal:
```
 PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND    
 6159 dbt       20   0 1466440 141092  98856 S   5.6  0.9   0:44.92 chromium-browse                                                                                    
31078 dbt       20   0  867744 142908  90000 S   2.0  0.9   4:50.11 chromium-browse                                                                                    
 9897 dbt       20   0  823340  28100  22540 S   0.7  0.2   0:00.06 soss          
 ......   
```
So the problem maybe result from soss-ros1 or rosmaster, then i look through the issues of rosmaster github page, find a similiar problem, then patch the [PR](https://github.com/ros/ros_comm/pull/1651/) to my local repository, rebuild  the revelant rosmaster code, The result is that it can solve the corresponding issue of rosmaster, but not this problem.

I will try to find reason for this, but for now, @mxgrey do you know this problem before or have any hint ?
Thanks.

**Comment by mxgrey** [comment]
I suspect it's most likely caused by some busy waiting in the soss plugin. We've seen busy waiting issues before and we have a strategy to fix it, but we haven't gotten around to implementing it yet.

I've gone ahead and opened an issue about it here: osrf/soss#24

I don't know for sure if that's the cause, but it would be my best guess right now.

**Comment by sparkskraps** [comment]
Yes, i have tested for that, that's the reason. Thank you :) 
Maybe in function SystemHandle::spin_once(), just call thread::sleep_for sleep some milliseconds just like other soss plugin to avoid do too heavy of a dead spin.

**Comment by mxgrey** [comment]
I'm concerned that the arbitrary sleep may have a negative impact on performance.

The `soss-websocket` plugin uses a sleep to reduce busy waiting, but that case is a bit different, because the websocket plugin has its own thread running asynchronously. The only thing that the sleep impacts for `soss-websocket` is how long it takes the plugin to respond to a SIGINT from the user. But if we were to introduce a sleep to `spin_once()` for `soss-ros1`, then the sleep would potentially impact all communication going to and from `soss-ros1`.

We could have `soss-ros1` split off its own thread for spinning and then change its `spin_once()` to be more like `soss-websocket`, but at that point we should really just go ahead and make all the changes proposed in osrf/soss#24.

If it's any consolation, you probably won't experience a noticeable performance hit from this busy waiting. It's true that you'll always see one core being fully utilized, but the operating system should have an easy time having that core interleave this busy waiting task with any other tasks that need to be processed, so it's not like the busy wait will stall anything else from getting executed. But let me know if you disagree with that, and you believe that this actually **is** having a noticeable negative impact on your computer's performance.

**Comment by sparkskraps** [comment]
Sorry to reply later, just busy with something.

Yes, you are right, we could not invoke sleep function at will, it will affect the commnuication related to soss-ros1.

I spent time browsing the soss-websocket source code. In it's spin_once function, as below.
Server.cpp:
```
bool spin_once() override
  {
    if(!_has_spun_once)
    {
      _has_spun_once = true;
      _server.start_accept();
    }

    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    // TODO(MXG): How do we know if the server is okay?
    return true;
  }
```
I could not enter into the start_accept() function.  But i think its function is to accept request from Client, So i think spliting a thread for soss-ros1 may be a good choice.  Then i roughly refer to soss-websocket for a simulation.
soss-ros1/packages/ros1/src/SystemHandle.cpp:
``` 
@@ -133,7 +137,13 @@ bool SystemHandle::okay() const
 //==============================================================================
 bool SystemHandle::spin_once()
 {
-  ros::spinOnce();
+  if(!has_spawn)
+  {
+       has_spawn = true;
+       ros::spinOnce();
+  }
+  std::this_thread::sleep_for(std::chrono::seconds(10));
+
   return ros::ok();
 }
```

It can solve higher cpu problem, and only affect the efficiency of responding to a SIGINT from the user. I think we just need to call ros::spin() once, right?
So, Is this a solution worth considering or we can design more resonable?

**Comment by azeey** [comment]
I'm not sure if is related, but I saw you mentioned [this PR](https://github.com/ros/ros_comm/pull/1651/) and thought the problem might be related to an issue we found a while ago: https://github.com/azeey/sandbox/tree/master/ros_high_cpu. In this case, the high CPU usage seems to occur when a `ros::NodeHandle` gets removed (goes out of scope) long before the node exits.

**Comment by sparkskraps** [comment]
> I'm not sure if is related, but I saw you mentioned [this PR](https://github.com/ros/ros_comm/pull/1651/) and thought the problem might be related to an issue we found a while ago: https://github.com/azeey/sandbox/tree/master/ros_high_cpu. In this case, the high CPU usage seems to occur when a `ros::NodeHandle` gets removed (goes out of scope) long before the node exits.

Thanks for your clue, I download your demo code and test，the phenoment just as you described, but i don't think it's related to this issue, and i believe this issue is result from the reason that @mxgrey has metioned above.

**Comment by fujitatomoya** [comment]
@friendsocket @azeey 

[this issue](https://github.com/ros/ros_comm/pull/1651/) is monotonic system time related issue, so it got nothing to do with this as far as I know. as you mentioned previously, this is related to wait timer to check if any events ready.

soss(ros2) side does not have this problem because of the following,

https://github.com/osrf/soss/blob/b531c2046e24684670a4a2ea2fd3c134fcba0591/packages/ros2/src/SystemHandle.cpp#L209-L213

but soss-ros1 does have this problem because of,

https://github.com/osrf/soss-ros1/blob/0a88d1126a6d5f453be55dff900ad93143258d8d/packages/ros1/src/SystemHandle.cpp#L134-L138

there are a few fixes i can think of,

- fix ros_comm spinOnce() with argument with ros::WallDuration() and specify timeout from soss-ros1.
- fix soss-ros1 with ros::WallDuration().sleep() for iteration loop.

maybe latter is better? what do you think?

**Comment by mxgrey** [comment]
> fix soss-ros1 with ros::WallDuration().sleep() for iteration loop.

Is there an advantage of using `ros::WallDuration().sleep()` instead of `std::this_thread::sleep_for()`?

If not, I'm worried about adding an arbitrary sleep duration to every cycle. Yes it would reduce the CPU usage, but it would introduce unnecessary delays.

In the `soss-ros2` implementation it won't wait that entire 100 millisecond duration unless there's no work to be done, so there's no harm there.

I think the solution I would recommend is to wait for the changes recommended in osrf/soss#24 so that the `soss-ros1` SystemHandle can just use `ros::spin()` and not worry about periodically waking up or waiting.

If the high CPU usage is causing specific issues for any current users, then I guess I'm open to adding some small amount of waiting to each spin cycle as a hold over until we can make the osrf/soss#24 changes, but I'm skeptical that the high CPU usage would cause worse problems than introducing an arbitrary wait time to every spin cycle.

**Comment by fujitatomoya** [comment]
@mxgrey 

actually i did not know https://github.com/osrf/soss/issues/24, and now i do agree with you. sorry for the noise.

**Comment by jamoralp** [comment]
Fixed in eProsima/ROS1-SH#29 :slightly_smiling_face: please reopen if the problem persists!
